src/AutoRest.CSharp/MgmtTest/Generation/ScenarioTestWriter.cs

```diff
@@ -0,0 +1,347 @@
+            this.testDef = testDef;
+            this.writer = writer;
+        }
+
+        public void WriteScenarioTest()
+        {
+            WriteUsings(writer);
+
+            using (writer.Namespace(TestNamespace))
+            {
+                writer.WriteXmlDocumentationSummary($"Test generated from Api Scenario {testDef.FilePath}");
+                writer.Append($"public partial class {TypeName:D} : ");
+                writer.Line($"{TestBaseName}");
+                using (writer.Scope())
+                {
+                    WriteTesterCtors();
+
+                    WritePrepareMethod();
+                    WriteCleanUpMethod();
+
+                    foreach (var scenario in testDef.Scenarios)
+                    {
+                        WriteScenarioMethod(scenario);
+                    }
+                }
+            }
+        }
+
+        protected void WriteUsings(CodeWriter writer)
+        {
+            writer.UseNamespace("System");
+            writer.UseNamespace("System.Threading.Tasks");
+            writer.UseNamespace("System.Net");
+            writer.UseNamespace("System.Collections.Generic");
+            writer.UseNamespace("Azure.Core.TestFramework");
+            writer.UseNamespace("Azure.ResourceManager.TestFramework");
+        }
+
+        protected void WritePrepareMethod() {
+
+        }
+
+        protected void WriteCleanUpMethod()
+        {
+
+        }
+
+        protected void FindResourceOrCollectionFromOperation(Operation operation, out Resource? resource, out ResourceCollection? collection, out MgmtClientOperation? clientOperation, out MgmtRestOperation? restOperation)
+        {
+            resource = null;
+            collection = null;
+            clientOperation = null;
+            restOperation = null;
+            foreach (var r in MgmtContext.Library.ArmResources) {
+                foreach (var co in r.AllOperations) {
+                    foreach (var ro in co) {
+                        if (ro.Operation == operation) {
+                            resource = r;
+                            clientOperation = co;
+                            restOperation = ro;
+                            return;
+                        }
+                    }
+                }
+            }
+            foreach (var c in MgmtContext.Library.ResourceCollections) {
+                foreach (var co in c.AllOperations) {
+                    foreach (var ro in co) {
+                        if (ro.Operation == operation) {
+                            collection = c;
+                            clientOperation = co;
+                            restOperation = ro;
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+
+        protected bool IsCollectionOperation(Operation operation, out ResourceCollection? collection, out MgmtClientOperation? clientOperation, out MgmtRestOperation? restOperation)
```

Comment:
Please consider if we could add the attribute `MaybeNullWhen(bool)` to the output variables. This helps the static analyzer for null checks and we could save a lot from the null checkings as well